File: src/main/java/com/aws/trading/ExchangeClientLatencyTestHandler.java
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: MIT-0
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this
 * software and associated documentation files (the "Software"), to deal in the Software
 * without restriction, including without limitation the rights to use, copy, modify,
 * merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package com.aws.trading;

import com.alibaba.fastjson2.JSON;
import com.alibaba.fastjson2.JSONObject;
import io.netty.buffer.ByteBuf;
import io.netty.buffer.Unpooled;
import io.netty.channel.Channel;
import io.netty.channel.ChannelHandlerContext;
import io.netty.channel.ChannelInboundHandlerAdapter;
import io.netty.channel.ChannelPromise;
import io.netty.handler.codec.http.FullHttpResponse;
import io.netty.handler.codec.http.HttpHeaders;
import io.netty.handler.codec.http.websocketx.*;
import io.netty.handler.ssl.SslContext;
import io.netty.handler.ssl.SslContextBuilder;
import io.netty.handler.ssl.util.InsecureTrustManagerFactory;
import io.netty.util.CharsetUtil;
import org.HdrHistogram.SingleWriterRecorder;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import javax.net.ssl.SSLException;
import java.net.URI;
import java.nio.charset.StandardCharsets;
import java.util.Random;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;

import static com.aws.trading.Config.COIN_PAIRS;
import static com.aws.trading.Config.USE_SSL;
import static com.aws.trading.RoundTripLatencyTester.printResults;

public class ExchangeClientLatencyTestHandler extends ChannelInboundHandlerAdapter {
    private static final Logger LOGGER = LogManager.getLogger(ExchangeClientLatencyTestHandler.class);
    private final WebSocketClientHandshaker handshaker;
    private final int apiToken;
    private final int test_size;
    public final URI uri;
    private final ExchangeProtocol protocol;

    private ChannelPromise handshakeFuture;
    private final ConcurrentHashMap<String, Long> orderSentTimeMap;
    private final ConcurrentHashMap<String, Long> cancelSentTimeMap;
    private long orderResponseCount = 0;
    private final SingleWriterRecorder hdrRecorderForAggregation;
    private long testStartTime = 0;
    private final Random random = new Random();

    public ExchangeClientLatencyTestHandler(ExchangeProtocol protocol, URI uri, int apiToken, int test_size) {
        this.uri = uri;
        this.protocol = protocol;
        var header = HttpHeaders.EMPTY_HEADERS;
        this.handshaker = WebSocketClientHandshakerFactory.newHandshaker(
                uri, WebSocketVersion.V13, null, false, header, 1280000);
        this.apiToken = apiToken;
        this.orderSentTimeMap = new ConcurrentHashMap<>(test_size);
        this.cancelSentTimeMap = new ConcurrentHashMap<>(test_size);
        this.test_size = test_size;
        this.hdrRecorderForAggregation = new SingleWriterRecorder(Long.MAX_VALUE, 2);
    }

    @Override
    public void handlerAdded(final ChannelHandlerContext ctx) throws Exception {
        this.handshakeFuture = ctx.newPromise();
    }

    @Override
    public void channelActive(final ChannelHandlerContext ctx) throws Exception {
        LOGGER.info("channel is active, starting websocket handshaking...");
        handshaker.handshake(ctx.channel());
    }

    @Override
    public void channelInactive(final ChannelHandlerContext ctx) throws Exception {
        LOGGER.info("Websocket client disconnected");
    }

    @Override
    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
        final Channel ch = ctx.channel();
        if (!handshaker.isHandshakeComplete()) {
            LOGGER.info("Websocket client is connected");
            var m = (FullHttpResponse) msg;
            handshaker.finishHandshake(ch, m);
            LOGGER.info("Websocket client is authenticating for {}", this.apiToken);
            //success, authenticate
            var channel = ctx.channel();
            channel.write(authMessage());
            channel.flush();
            handshakeFuture.setSuccess();
            return;
        }

        if (msg instanceof FullHttpResponse) {
            final FullHttpResponse response = (FullHttpResponse) msg;
            throw new Exception("Unexpected FullHttpResponse (getStatus=" + response.getStatus() + ", content="
                    + response.content().toString(CharsetUtil.UTF_8) + ')');
        }
        final WebSocketFrame frame = (WebSocketFrame) msg;
        if (frame instanceof TextWebSocketFrame) {
            this.onTextWebSocketFrame(ctx, (TextWebSocketFrame) frame);
        } else if (frame instanceof PongWebSocketFrame) {
        } else if (frame instanceof CloseWebSocketFrame) {
            LOGGER.info("received CloseWebSocketFrame, closing the channel");
            ch.close();
        } else if (frame instanceof BinaryWebSocketFrame) {
            LOGGER.info(frame.content().toString());
        }

    }

    private TextWebSocketFrame subscribeMessage() {
        return new TextWebSocketFrame(Unpooled.wrappedBuffer(ExchangeProtocolImpl.SUBSCRIBE_MSG));
    }

    private TextWebSocketFrame authMessage() {
        return new TextWebSocketFrame(Unpooled.wrappedBuffer(
                ExchangeProtocolImpl.AUTH_MSG_HEADER,
                Integer.toString(this.apiToken).getBytes(StandardCharsets.UTF_8),
                ExchangeProtocolImpl.MSG_END)
        );
    }

    @Override
    public void exceptionCaught(final ChannelHandlerContext ctx, final Throwable cause) throws Exception {
        LOGGER.error(cause);
        if (!handshakeFuture.isDone()) {
            handshakeFuture.setFailure(cause);
        }
        ctx.close();
    }

    public ChannelPromise getHandshakeFuture() {
        return handshakeFuture;
    }

    private void onTextWebSocketFrame(ChannelHandlerContext ctx, TextWebSocketFrame textFrame) throws InterruptedException {
        long eventReceiveTime = System.nanoTime();
        ByteBuf buf = textFrame.content();

        final byte[] bytes;
        int offset = 0;
        final int length = buf.readableBytes();
        if (buf.hasArray()) {
            bytes = buf.array();
            offset = buf.arrayOffset();
        } else {
            bytes = new byte[length];
            buf.getBytes(buf.readerIndex(), bytes);
        }
        buf.clear();
        buf.release();

        JSONObject parsedObject = JSON.parseObject(bytes, offset, bytes.length - offset, StandardCharsets.UTF_8);
        Object type = parsedObject.getString("type");

        if ("BOOKED".equals(type) || type.equals("DONE")) {
            //LOGGER.info("eventTime: {}, received ACK: {}",eventReceiveTime, parsedObject);
            String clientId = parsedObject.getString("client_id");
            if (type.equals("BOOKED")) {
                if (calculateRoundTrip(eventReceiveTime, clientId, orderSentTimeMap)) return;
                var pair = parsedObject.getString("instrument_code");
                sendCancelOrder(ctx, clientId, pair);
            } else {
                if (calculateRoundTrip(eventReceiveTime, clientId, cancelSentTimeMap)) return;
                sendOrder(ctx);
            }
            if (orderResponseCount % test_size == 0) {
                printResults(hdrRecorderForAggregation, test_size);
            }
        } else if ("AUTHENTICATED".equals(type)) {
            LOGGER.info("{}", parsedObject);
            ctx.channel().writeAndFlush(subscribeMessage());
        } else if ("SUBSCRIPTIONS".equals(type)) {
            LOGGER.info("{}", parsedObject);
            this.testStartTime = System.nanoTime();
            sendOrder(ctx);
        } else {
            LOGGER.error("Unhandled object {}", parsedObject);
        }
    }

    private void sendCancelOrder(ChannelHandlerContext ctx, String clientId, String pair) {
        TextWebSocketFrame cancelOrder = protocol.createCancelOrder(pair, clientId);
        //LOGGER.info("Sending cancel order seq: {}, order: {}", sequence, cancelOrder.toString(StandardCharsets.UTF_8));
        try {
            ctx.channel().write(cancelOrder, ctx.channel().voidPromise()).await();
        } catch (InterruptedException e) {
            LOGGER.error(e);
        }
        var cancelSentTime = System.nanoTime();
        //LOGGER.info("cancel sent time for clientId: {} - {}",clientId, cancelSentTime);
        this.cancelSentTimeMap.put(clientId, cancelSentTime);
        ctx.channel().flush();
        orderResponseCount += 1;
    }

    private boolean calculateRoundTrip(long eventReceiveTime, String clientId, ConcurrentHashMap<String, Long> cancelSentTimeMap) {
        long roundTripTime;
        Long cancelSentTime = cancelSentTimeMap.remove(clientId);
        if (null == cancelSentTime || eventReceiveTime < cancelSentTime) {
            LOGGER.error("no order sent time found for order {}", clientId);
            return true;
        }
        roundTripTime = eventReceiveTime - cancelSentTime;
        //LOGGER.info("round trip time for client id {}: {} = {} - {}", clientId, roundTripTime, eventReceiveTime, cancelSentTime);
        if (roundTripTime > 0) {
            //LOGGER.info("recording round trip time");
            hdrRecorderForAggregation.recordValue(roundTripTime);
        }
        return false;
    }

    void sendOrder(ChannelHandlerContext ch) throws InterruptedException {

        var pair = COIN_PAIRS.get(random.nextInt(COIN_PAIRS.size()));
        var clientId = UUID.randomUUID().toString();
        var order = protocol.createBuyOrder(pair, clientId);
        ch.write(order, ch.voidPromise()).await();
        var time = System.nanoTime();
        //LOGGER.info("sending order: {}, time: {}", clientId, time);
        orderSentTimeMap.put(clientId, time);
        //LOGGER.info("sending pair, clientId: {}, {}", pair, clientId);
        ch.flush();
        orderResponseCount += 1;
    }
}

File: hft_client/cpp-client/ExchangeClientLatencyTestHandler.cpp
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: MIT-0
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this
 * software and associated documentation files (the "Software"), to deal in the Software
 * without restriction, including without limitation the rights to use, copy, modify,
 * merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
#include "ExchangeProtocol.h"
#include "ExchangeClientLatencyTestHandler.h"
#include "Logger.h"
#include "Config.h"
#include <string_view>
#include <memory>
#include <nlohmann/json.hpp>
#include <iostream>
#include <unordered_map>
#include <openssl/rand.h>
//#include <hdr_histogram.h>
#include "hdr/hdr_histogram.h"
#include "hdr/hdr_histogram_log.h"
#include <sys/stat.h>
#include <algorithm>
using namespace std;
using json = nlohmann::json;

ExchangeClientLatencyTestHandler::ExchangeClientLatencyTestHandler(int apiToken, const std::string& uri)
        : m_apiToken(apiToken),
        m_uri(uri),
        testSize(Config::TEST_SIZE),
        orderSentTimeMap(Config::TEST_SIZE),
        cancelSentTimeMap(Config::TEST_SIZE)
        {
    this->histogram = new hdr_histogram();
    hdr_init(
            1,  // Minimum value
            INT64_C(3600000000000),  // Maximum value (1 hour in nanoseconds)
            3,  // Number of significant figures
            &histogram);  // Pointer to initialise
}

ExchangeClientLatencyTestHandler::~ExchangeClientLatencyTestHandler() {
    if (histogram) {
        hdr_close(histogram);
    }
}

string ExchangeClientLatencyTestHandler::authMessage() {
    // Replace with your authentication message implementation
    return ExchangeProtocol::AUTH_MSG_HEADER + std::to_string(m_apiToken) + ExchangeProtocol::MSG_END;
}


std::string uuid_v4_gen()
{
    union
    {
        struct
        {
            uint32_t time_low;
            uint16_t time_mid;
            uint16_t time_hi_and_version;
            uint8_t  clk_seq_hi_res;
            uint8_t  clk_seq_low;
            uint8_t  node[6];
        };
        uint8_t __rnd[16];
    } uuid;

    RAND_bytes(uuid.__rnd, sizeof(uuid));
    // Refer Section 4.2 of RFC-4122
    // https://tools.ietf.org/html/rfc4122#section-4.2
    uuid.clk_seq_hi_res = (uint8_t) ((uuid.clk_seq_hi_res & 0x3F) | 0x80);
    uuid.time_hi_and_version = (uint16_t) ((uuid.time_hi_and_version & 0x0FFF) | 0x4000);

    char buffer[37];
    snprintf(buffer, sizeof(buffer), "%08x-%04x-%04x-%02x%02x-%02x%02x%02x%02x%02x%02x",
             uuid.time_low, uuid.time_mid, uuid.time_hi_and_version,
             uuid.clk_seq_hi_res, uuid.clk_seq_low,
             uuid.node[0], uuid.node[1], uuid.node[2],
             uuid.node[3], uuid.node[4], uuid.node[5]);

    return std::string(buffer);
}

bool ExchangeClientLatencyTestHandler::calculateRoundTrip(chrono::steady_clock::time_point eventReceiveTime, string_view clientId, unordered_map<string, chrono::steady_clock::time_point>& sentTimeMap) {
    auto it = sentTimeMap.find(string(clientId));
    if (it == sentTimeMap.end() || eventReceiveTime < it->second) {
        logger("no order sent time found for order " + std::string(clientId));
        return true;
    }
    auto roundTripTime = eventReceiveTime - it->second;
    if (roundTripTime.count() > 0) {
        long long roundTripTimeInNanoseconds = chrono::duration_cast<chrono::nanoseconds>(roundTripTime).count();
        hdr_record_value(
                histogram,  // Histogram to record to
                roundTripTimeInNanoseconds);
    }
    sentTimeMap.erase(it);
    return false;
}


void ExchangeClientLatencyTestHandler::hdrPrint(){
    // Calculate and store the percentile values
    int64_t value50 = hdr_value_at_percentile(histogram, 50.0);
    int64_t value90 = hdr_value_at_percentile(histogram, 90.0);
    int64_t value95 = hdr_value_at_percentile(histogram, 95.0);
    int64_t value99 = hdr_value_at_percentile(histogram, 99.0);
    int64_t value99_9 = hdr_value_at_percentile(histogram, 99.9);
    int64_t value99_99 = hdr_value_at_percentile(histogram, 99.99);
    int64_t valueW = hdr_max(histogram);
    std::stringstream ss;
    ss << "Percentiles: {" << std::endl;
    ss << "        \"50.0%\":\"" << value50 << "ns\"," << std::endl;
    ss << "        \"90.0%\":\"" << value90 << "ns\"," << std::endl;
    ss << "        \"95.0%\":\"" << value95 << "ns\"," << std::endl;
    ss << "        \"99.0%\":\"" << value99 << "ns\"," << std::endl;
    ss << "        \"99.9%\":\"" << value99_9 << "ns\"," << std::endl;
    ss << "        \"99.99%\":\"" << value99_99 << "ns\"," << std::endl;
    ss << "        \"W\":\"" << valueW << "ns\"" << std::endl;
    ss << "}";
    std::cout << ss.str() << std::endl;

    // Save histogram to .hlog file in host-named directory
    saveHistogramToFile();
    
    // Reset histogram for next interval (like Java's hdr.reset())
    hdr_reset(histogram);
}
void ExchangeClientLatencyTestHandler::saveHistogramToFile() {
    // Create host-named directory (dots replaced with underscores)
    std::string folder = Config::HOST;
    std::replace(folder.begin(), folder.end(), '.', '_');
    mkdir(folder.c_str(), 0755);

    std::string path = folder + "/histogram_cpp.hlog";
    FILE* f = fopen(path.c_str(), "a");
    if (!f) {
        logger("Failed to open histogram log file: " + path);
        return;
    }

    struct hdr_log_writer writer;
    hdr_log_writer_init(&writer);

    struct timespec now;
    clock_gettime(CLOCK_REALTIME, &now);

    hdr_log_write_header(&writer, f, "C++ HFT Client 0.0.1", &now);
    hdr_log_write(&writer, f, &now, &now, histogram);
    fclose(f);

    logger("Histogram saved to " + path);
}

websocketpp::connection_hdl ExchangeClientLatencyTestHandler::get_hdl() const{
    return this->hdl;
}

template<typename T>
void ExchangeClientLatencyTestHandler::sendOrder(T* c, websocketpp::connection_hdl m_hdl) {
//    static random_device rd;
//    static mt19937 gen(rd());
//    static uniform_int_distribution<> dis(0, Config::COIN_PAIRS.size() - 1);
//    string pair = Config::COIN_PAIRS[dis(gen)];
    string_view pair = Config::COIN_PAIRS[0];
    auto clientId = uuid_v4_gen();
    string order = protocol->createBuyOrder(pair, clientId);
    // std::cout << order << std::endl;
    websocketpp::lib::error_code ec;
    c->send(m_hdl, order,websocketpp::frame::opcode::text, ec);
    auto time = chrono::steady_clock::now();
    orderSentTimeMap.insert(std::make_pair(string(clientId), time));
    orderResponseCount += 1;
}

template<typename T>
void ExchangeClientLatencyTestHandler::sendCancelOrder(T* c, websocketpp::connection_hdl m_hdl, string_view clientId, string_view pair) {
    string cancelOrder = protocol->createCancelOrder(pair, clientId);
    // std::cout << cancelOrder << std::endl;
    try {
        websocketpp::lib::error_code ec;
        c->send(m_hdl, cancelOrder, websocketpp::frame::opcode::text, ec);
        // std::cout << "cancel is sent" << std::endl;
    } catch (const exception& e) {
        logger(e.what());
    }
    auto cancelSentTime = chrono::steady_clock::now();
    //LOGGER.info("cancel sent time for clientId: {} - {}",clientId, cancelSentTime);
    cancelSentTimeMap.insert(std::make_pair(string(clientId), cancelSentTime));
    orderResponseCount += 1;
}

template<typename T>
void ExchangeClientLatencyTestHandler::on_message(T* c, websocketpp::connection_hdl m_hdl, typename T::message_ptr msg) {
    auto eventReceiveTime = chrono::steady_clock::now();

    auto parsedObject = json::parse(msg->get_payload());
    string_view type = parsedObject["type"].template get<string_view>();


    if (type == "BOOKED" || type == "DONE") {
        // logger("Order ack received");
        string_view clientId = parsedObject["client_id"].template get<string_view>();
        if (type == "BOOKED") {
            if (calculateRoundTrip(eventReceiveTime, clientId, orderSentTimeMap)) return;
            string_view pair = parsedObject["instrument_code"].template get<string_view>();
            sendCancelOrder(c, m_hdl, clientId, pair);
        } else {
            if (calculateRoundTrip(eventReceiveTime, clientId, cancelSentTimeMap)) return;
            sendOrder(c, m_hdl);
        }
        if (orderResponseCount % Config::TEST_SIZE == 0) {
            hdrPrint();

            // Test complete — close connection to exit gracefully
            logger("Test completed. Reached TEST_SIZE: " + std::to_string(Config::TEST_SIZE) + ". Exiting gracefully.");
            websocketpp::lib::error_code ec;
            c->close(m_hdl, websocketpp::close::status::normal, "test complete", ec);
            return;
        }
    } else if (type == "AUTHENTICATED") {
        logger(parsedObject.dump());
        websocketpp::lib::error_code ec;
        c->send(m_hdl, ExchangeProtocol::SUBSCRIBE_MSG,websocketpp::frame::opcode::text, ec);
    } else if (type == "SUBSCRIPTIONS") {
        logger(parsedObject.dump());
        this->testStartTime = chrono::steady_clock::now();
        sendOrder(c, m_hdl);
    } else {
        logger("Unhandled object " + parsedObject.dump());
    }
}

template<typename T>
void ExchangeClientLatencyTestHandler::on_open(T* c, websocketpp::connection_hdl m_hdl) {
    logger("WebSocket client is connected");
    websocketpp::lib::error_code ec;
    c->send(m_hdl, authMessage(), websocketpp::frame::opcode::text, ec);
    if (ec) {
        logger("Error sending message: " + ec.message());
    } else {
        logger("WebSocket client is authenticating for " + std::to_string(m_apiToken));
    }
}

template<typename T>
void ExchangeClientLatencyTestHandler::on_close(T* c, websocketpp::connection_hdl hdl) {
    logger("Connection closed");
}

template void ExchangeClientLatencyTestHandler::sendOrder<ExchangeClientLatencyTestHandler::ssl_client>(ssl_client* c, websocketpp::connection_hdl hdl);
template void ExchangeClientLatencyTestHandler::sendOrder<ExchangeClientLatencyTestHandler::non_ssl_client>(non_ssl_client* c, websocketpp::connection_hdl hdl);
template void ExchangeClientLatencyTestHandler::sendCancelOrder<ExchangeClientLatencyTestHandler::ssl_client>(ssl_client* c, websocketpp::connection_hdl hdl, string_view clientId, string_view pair);
template void ExchangeClientLatencyTestHandler::sendCancelOrder<ExchangeClientLatencyTestHandler::non_ssl_client>(non_ssl_client* c, websocketpp::connection_hdl hdl, string_view clientId, string_view pair);
template void ExchangeClientLatencyTestHandler::on_open<ExchangeClientLatencyTestHandler::ssl_client>(ssl_client* c, websocketpp::connection_hdl hdl);
template void ExchangeClientLatencyTestHandler::on_open<ExchangeClientLatencyTestHandler::non_ssl_client>(non_ssl_client* c, websocketpp::connection_hdl hdl);
template void ExchangeClientLatencyTestHandler::on_message<ExchangeClientLatencyTestHandler::ssl_client>(ssl_client* c, websocketpp::connection_hdl hdl, ssl_client::message_ptr msg);
template void ExchangeClientLatencyTestHandler::on_message<ExchangeClientLatencyTestHandler::non_ssl_client>(non_ssl_client* c, websocketpp::connection_hdl hdl, non_ssl_client::message_ptr msg);
template void ExchangeClientLatencyTestHandler::on_close<ExchangeClientLatencyTestHandler::ssl_client>(ssl_client* c, websocketpp::connection_hdl hdl);
template void ExchangeClientLatencyTestHandler::on_close<ExchangeClientLatencyTestHandler::non_ssl_client>(non_ssl_client* c, websocketpp::connection_hdl hdl);
